[MXNet-1334][Fit API]base class for estimator and eventhandler

roywei opened the pull request:
## Description ##
This is the first PR for the Gluon Fit API proposed here:  [Gluon Fit API tech design](https://cwiki.apache.org/confluence/display/MXNET/Gluon+Fit+API+-+Tech+Design)

It add base classes for estimator and eventhandlers, and an example.  We would like to merge into fit-api branch so other controbutors can work on top of it before making it available on master branch.

JIRA epic can be found here: https://issues.apache.org/jira/browse/MXNET-1333
We are separating this into a few PRs, all tracked by JIRA issues under the epic.

We will open follow up PRs for the following tasks:
1.  CNN and RNN examples using the fit API https://github.com/apache/incubator-mxnet/pull/14350
2. User experience improvement on initialization, context management, and hybridization
3. More eventhandlers like Checkpoint, early stopping. 
4. Multi output support for Multi-task and SSD use cases and examples
5. Multi trainer support for encoder decoder and neural machine translation example.
6. Unit tests and integration tests 
7. Documentation 



## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-1334](https://issues.apache.org/jira/browse/MXNET-1334) created
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
1. Base class for estimator with fit method
2. Base class for event handlers
3. simple cnn example on mnist dataset

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

@ankkhedia

vandanavk commented:
@mxnet-label-bot add [Gluon, pr-work-in-progress]

ptrendx commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+                metric.reset()
+
+            for i, batch in enumerate(train_data):
+                data, label = self._batch_fn(batch, self.context)
```
Hi @roywei, in NVIDIA DALI we return already sliced data, could you make `_batch_fn` optional?

roywei replied:
Hi @ptrendx sounds good！any examples on multi-gpu? would love to check it out.

nswamy commented on example/gluon/estimator_example/mnist_cnn.py:
```diff
@@ -0,0 +1,67 @@
+    mnist_train = data.vision.MNIST(root=root, train=True)
+    mnist_test = data.vision.MNIST(root=root, train=False)
+    num_workers = 0 if sys.platform.startswith('win32') else 4
+    train_iter = data.DataLoader(
```
nit: rename to train_dataloader

roywei replied:
removed this example will be covered in integration tests

nswamy commented on example/gluon/estimator_example/mnist_cnn.py:
```diff
@@ -0,0 +1,67 @@
+    train_iter = data.DataLoader(
+        mnist_train.transform_first(transformer), batch_size, shuffle=True,
+        num_workers=num_workers)
+    test_iter = data.DataLoader(
```
nit: rename

roywei replied:
removed this example will be covered in integration tests

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+            self.loss = [loss]
+        else:
+            self.loss = loss or []
+        if isinstance(metrics, EvalMetric):
```
this does not work for a list of metrics?
```
>>> class A:
...     def __init(self):
...             pass
...
>>> A
<class __main__.A at 0x721d0fe88>
>>> x = A
>>> x = [A(), A()]
>>> isinstance(x, A)
False
```

roywei replied:
yes, it accepts single metric object or  list of metrics(in the else part)

nswamy replied:
regardless of whether a single metric or a list of them are passed, we should have this validation

roywei replied:
@nswamy updated

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+        self.train_stats['epochs'] = []
+        self.train_stats['learning_rate'] = []
+        # time used for each epoch
+        self.train_stats['step'] = ''
```
would `time_per_epoch` or `time_epoch` be better name for the key?

roywei replied:
wrong comment, updated

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+
+        self.initializer = initializer
+        # store training statistics
+        self.train_stats = {}
```
can you want to pull initializing stats into a small helper method ? `_init_stats` or something

roywei replied:
going to put train_stats as an object and add related methods in a follow up PR.

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+            if num_gpus() > 0:
+                # only use 1 GPU by default
+                if num_gpus() > 1:
+                    warnings.warn("You have multiple GPUs, gpu(0) will be used by default."
```
we should use all the GPUs on the system?

roywei replied:
This behavior is borrowed from keras, if no context specified, use 1 gpu by default. Since for beginner use cases, running small dataset over multiple GPU will cause communication overhead.

piyushghai replied:
Can using the number of gpus be a parameter that the user can pass to Estimator ?

roywei replied:
@piyushghai Agree that using gpus will be simpler. But the context arg is exposed many places in MXNet/Gluon, from inference to traninig. right now there is no good way to only allow gpu as there are too many api accepting context argument. User may get confused and error if we have both gpus and context.
e.g.
```
net.initialize(..., ctx=[mx.gpu(0)])
est = Estimator( net, ..., gpus=2)
```

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+        if self.initializer:
+            if self._is_initialized():
+                # if already initialized, re-init with user specified initializer
+                warnings.warn("You have already initialized your net, it will be forced re-initialized "
```
'Network is initialized, re-initializing with  {}'.format(initializer.name)`

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+                self.context = [cpu()]
+
+        # initialize the network
+        if self.initializer:
```
suggest to move the initializing and checking of initializing into one method.

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+        label = gluon.utils.split_and_load(batch[1], ctx_list=ctx, batch_axis=0)
+        return data, label
+
+    def fit(self, train_data,
```
nit: `train_data`-> `train_dataloader`, ..
if `valid_data` is not passed are you going to split the train data with some percentage.  if not i think we should

roywei replied:
currently dataloader does not support split into train and val after it's created. User can create that split outside estimator using a [random sampler ](https://mxnet.incubator.apache.org/versions/1.0.0/api/python/gluon/data.html#mxnet.gluon.data.DataLoader).  Same behavior with [pytorch](https://stackoverflow.com/questions/50544730/how-do-i-split-a-custom-dataset-into-training-and-test-datasets) . Keras allows this but only because the fit method is accepting numpy arrays directly and it can be slow.

nswamy replied:
`val_data` is missing in the signature? Also it should not be optional if you expect user to pass?

nswamy replied:
please rename to `train_dataloader`

nswamy commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+            batch_size=None,
+            event_handlers=None):
+
+        if not batch_size:
```
log.info( with the batch size you are using).

roywei replied:
will be added in train_begin in loggin handler with the change on train_stats in a follow up PR

nswamy commented on python/mxnet/gluon/estimator/event_handler.py:
```diff
@@ -0,0 +1,100 @@
+            self.logger.addHandler(filehandler)
+
+    def train_begin(self):
+        pass
```
might be a good idea to log all the hyper params and defaults used here

roywei replied:
will be added with the change on train_stats in a follow up PR

nswamy commented on python/mxnet/gluon/estimator/event_handler.py:
```diff
@@ -0,0 +1,100 @@
+        pass
+        # logger.info(opt)
+
+    def train_end(self):
```
output informational message like number of epochs run, train_loss and valid_loss , other metrics at the end, etc.,

roywei replied:
will be added with the change on train_stats in a follow up PR

ankkhedia commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+            self.trainers = [trainers]
+        else:
+            self.trainers = trainers or []
+        if not self.trainers:
```
Are we dealing with multiple trainer case over here (e.g. Multi task classification)

roywei replied:
will be addressed in another PR (TODO item 4)

ankkhedia commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,203 @@
+                    handler.batch_end()
+
+            for metric in self.metrics + self.loss_metrics:
+                self.train_stats['train_' + metric.name].append(metric.get()[1])
```
Validation metrics would also be needed in train_stats. How are we dealing with it?

roywei replied:
It's tracked [here](https://issues.apache.org/jira/browse/MXNET-1349), will be address in follow up PR

roywei commented on example/gluon/estimator_example/mnist_cnn.py:
```diff
@@ -0,0 +1,67 @@
+        transformer += [data.vision.transforms.Resize(resize)]
+    transformer += [data.vision.transforms.ToTensor()]
+    transformer = data.vision.transforms.Compose(transformer)
+    mnist_train = data.vision.MNIST(root=root, train=True)
```
removed this example will be covered in integration tests

roywei commented:
@nswamy @ankkhedia I have addressed comments, added checkpoint and early stopping handler, could you take another look? Thanks

Remaining todo list from your comments, will be addressed in follow up PRs
 1) update train stats (change to object instead of dictionary), initialization logic [MXNet-1340](https://issues.apache.org/jira/browse/MXNET-1340)
2. Validation logic [MXNet-1349](https://issues.apache.org/jira/browse/MXNET-1349)
3. multi trainer [MXNet-1339](https://issues.apache.org/jira/browse/MXNET-1339)

roywei commented:
@mxnet-label-bot update[Gluon, pr-awaiting-review]

piyushghai commented on python/mxnet/gluon/estimator/estimator.py:
```diff
@@ -0,0 +1,267 @@
+                    elif isinstance(train_data, DataIter):
+                        data, label = self._batch_fn(batch, self.context, is_iterator=True)
+                    else:
+                        raise ValueError("You are using a custom iteration, please also provide "
```
Nit : custom ```iterator```

piyushghai commented on tests/python/unittest/test_gluon_event_handler.py:
```diff
@@ -0,0 +1,92 @@
+    early_stopping = [event_handler.EarlyStoppingHandler(est, monitor,
+                                                         patience=patience,
+                                                          mode=mode)]
+    est.fit(test_data, event_handlers=early_stopping, epochs=1)
```
No ```assertions``` here ?

piyushghai commented on python/mxnet/gluon/estimator/event_handler.py:
```diff
@@ -0,0 +1,307 @@
+        epoch = self._estimator.train_stats['epochs'][-1]
+        msg = '\n[Epoch %d] finished in %.3fs: ' % (epoch, epoch_time)
+        for key in self._estimator.train_stats.keys():
+            if key.startswith('train_') or key.startswith('test_'):
```
key should be ```val_``` ?

nswamy commented on tests/python/unittest/test_gluon_event_handler.py:
```diff
@@ -0,0 +1,92 @@
+    est = estimator.Estimator(net, loss=ce_loss, metrics=acc)
+    logging_handler = [event_handler.LoggingHandler(est, file_name=file_name, file_location=tmpdir)]
+    est.fit(test_data, event_handlers=logging_handler, epochs=1)
+    assert os.path.isfile(output_dir)
```
do you want to do a sanity check that validation accuracy for last epoch is logged(also that training process completed?

nswamy commented on python/mxnet/gluon/estimator/event_handler.py:
```diff
@@ -0,0 +1,307 @@
+        batch_time = time.time() - self.batch_start
+        epoch = self._estimator.train_stats['epochs'][-1]
+        step = self._estimator.train_stats['step']
+        msg = '[Epoch %d] [Step %s] time/step: %.3fs ' % (epoch, step, batch_time)
```
I don't think we use the term `step`, lets not introduce new terms lets continue to use `batch`, look at existing logging

nswamy commented on python/mxnet/gluon/estimator/event_handler.py:
```diff
@@ -0,0 +1,307 @@
+                self._estimator.net.save_parameters(self.filepath)
+
+
+class EarlyStoppingHandler(EventHandler):
```
Can you please see if we can have a integration test for this?

nswamy commented:
I will merge this code since its on a branch, please make sure to address all comments in subsequent PRs before a PR to the master
